docker_cli/_shell_cmds.py

```diff
@@ -0,0 +1,369 @@
+    \t  -   :func:`~docker_cli.PackageManager.generate_configure_command`
+    -   Returning an associated name and filetype
+    """
+
+    @staticmethod
+    @abstractmethod
+    def generate_install_command(
+        targets: Iterable[str], *, stringify: bool = False, clean: bool = True
+    ) -> Union[List[str], str]:
+        """
+        Returns an set of commands to install the targets.
+
+        Parameters
+        ----------
+        targets : Sequence[str]
+            The install targets.
+        stringify : bool, optional
+            Whether to return a string or list - if true, will return a string.
+            Defaults to False.
+        clean : bool, optional
+            Whether or not to add a clean command to the list.
+            Defaults to True.
+
+        Returns
+        -------
+        Union[List[str], str]
+            The commands.
+        """
+        ...
+
+    @staticmethod
+    @abstractmethod
+    def generate_package_command(
+        target: str, *, stringify: bool = False
+    ) -> Union[List[str], str]:
+        """
+        Returns a command to install a single package.
+
+        Uses the package tool underlying the package manager if one is
+        available. (e.g. dpkg or rpm)
+
+        Parameters
+        ----------
+        target : str
+            The install target.
+        stringify : bool, optional
+            Whether to return a string or list - if true, will return a string.
+            Defaults to False.
+
+        Returns
+        -------
+        Union[List[str], str]
+            The commands.
+        """
+        ...
+
+    @staticmethod
+    @abstractmethod
+    def generate_configure_command(*, stringify: bool = False) -> Union[List[str], str]:
+        """
+        Returns commands to configure and update the package manager.
+
+        Parameters
+        ----------
+        stringify : bool, optional
+            Whether to return a string or list - if true, will return a string.
+            Defaults to False.
+
+        Returns
+        -------
+        Union[List[str], str]
+            The commands.
+        """
+        ...
+
+    @property
+    @abstractmethod
+    def name(self):
+        """
+        The package manager's command name.
+        """
+        ...
+
+    @property
+    @abstractmethod
+    def file_type(self):
+        """
+        The filetype associated with this package manager.
+        """
+        ...
+
+
+class Yum(PackageManager):
+    @staticmethod
+    def generate_install_command(
+        targets: Iterable[str],
+        *,
+        stringify: bool = False,
+        clean: bool = True,
+    ) -> Union[List[str], str]:
+        retval = []
+        retval += ["yum", "install", "-y"]
+        retval.extend(targets)
+        if clean:
+            retval += ["&&", "yum", "clean", "all"]
+            retval += ["&&", "rm", "-rf", "/var/cache/yum"]
+        if stringify:
+            return " ".join(retval)
```

Comment:
Let's use [`shlex.join()`](https://docs.python.org/3/library/shlex.html#shlex.join) for this instead

```suggestion
            return shlex.join(retval)
```

Reply:
Done for some and not for others; `shlex.join()` actually breaks some of these. Basically it looks like `shlex.join()` wants to create only a single token for the python shell. The result of this is that `&&` is changed by the function into `'&&'` when it's joined, which breaks the command.

The options here I think are to either break this command up into its subsequent tokens and `shlex.join()` them followed by a `' && '.join()` of the tokens into a single command, or to go with the `" ".join()` way currently in use for any command that has `&&` in it.

Reply:
Oh I see. Thanks for the heads up.

Well I guess I still think that the simplest solution is probably just to remove the `stringify` option. In that case, you can just always return a string with the proper formatting -- no need to use either `shlex.join()` or `" ".join()`.